diff --git a/src/display.rs b/lzgl-0.6.2/src/display.rs
index 8e8f510..a27b1dc 100644
--- a/src/display.rs
+++ b/src/display.rs
@@ -20,12 +20,12 @@ type Result<T> = result::Result<T, DisplayError>;
 
 /// An LVGL-registered display. Equivalent to an `lv_disp_t`.
 pub struct Display {
-    pub(crate) disp: NonNull<lvgl_sys::lv_disp_t>,
+    pub disp: NonNull<lvgl_sys::lv_disp_t>,
     drop: Option<unsafe extern "C" fn()>,
 }
 
 impl<'a> Display {
-    pub(crate) fn from_raw(
+    pub fn from_raw(
         disp: NonNull<lvgl_sys::lv_disp_t>,
         drop: Option<unsafe extern "C" fn()>,
     ) -> Self {
diff --git a/src/widgets/label.rs b/lzgl-0.6.2/src/widgets/label.rs
index 7d4e08d..5bb8652 100644
--- a/src/widgets/label.rs
+++ b/src/widgets/label.rs
@@ -4,7 +4,7 @@
 mod alloc_imp {
     use crate::widgets::Label;
     //use crate::LvError;
-    use cstr_core::CString;
+    use alloc::ffi::CString;
     //use core::convert::TryFrom;
 
     impl<S: AsRef<str>> From<S> for Label {
